Source/CarthageKit/BackTrackingResolver.swift

```diff
@@ -0,0 +1,512 @@
+		let dependencyRetriever = DependencyRetriever(versionsForDependency: versionsForDependency,
+													  dependenciesForDependency: dependenciesForDependency,
+													  resolvedGitReference: resolvedGitReference,
+													  pinnedVersions: pinnedVersions)
+		let updatableDependencyNames = dependenciesToUpdate.map { Set($0) } ?? Set()
+		let requiredDependencies: AnySequence<DependencyEntry>
+		let hasSpecificDepedenciesToUpdate = !updatableDependencyNames.isEmpty
+
+		if hasSpecificDepedenciesToUpdate {
+			requiredDependencies = AnySequence(dependencies.filter { dependency, _ in
+				updatableDependencyNames.contains(dependency.name) || pinnedVersions[dependency] != nil
+			})
+		} else {
+			requiredDependencies = AnySequence(dependencies)
+		}
+
+		do {
+			let dependencySet = try DependencySet(requiredDependencies: requiredDependencies,
+												  updatableDependencyNames: updatableDependencyNames,
+												  retriever: dependencyRetriever)
+			let resolverResult = try backtrack(dependencySet: dependencySet)
+
+			switch resolverResult.state {
+			case .accepted:
+				break
+			case .rejected:
+				if let rejectionError = dependencySet.rejectionError {
+					throw rejectionError
+				} else {
+					throw CarthageError.unresolvedDependencies(dependencySet.unresolvedDependencies.map { $0.name })
+				}
+			}
+
+			result = .success(resolverResult.dependencySet.resolvedDependencies)
+		} catch let error {
+			let carthageError: CarthageError = (error as? CarthageError) ?? CarthageError.internalError(description: error.localizedDescription)
+
+			result = .failure(carthageError)
+		}
+
+		return SignalProducer(result: result)
+	}
+
+	/**
+	Recursive backtracking algorithm to resolve the dependency set.
+	
+	See: https://en.wikipedia.org/wiki/Backtracking
+	*/
+	private func backtrack(dependencySet: DependencySet) throws -> (state: ResolverState, dependencySet: DependencySet) {
+		if dependencySet.isRejected {
+			return (.rejected, dependencySet)
+		} else if dependencySet.isComplete {
+			return (.accepted, dependencySet)
+		}
+
+		var result: (state: ResolverState, dependencySet: DependencySet)? = nil
+		while result == nil {
+			// Use an autorelease pool here to keep memory usage down
+			try autoreleasepool {
+				// Keep iterating until there are no subsets to resolve anymore
+				if let subSet = try dependencySet.popSubSet() {
+					// Backtrack again with this subset
+					let nestedResult = try backtrack(dependencySet: subSet)
+
+					switch nestedResult.state {
+					case .rejected:
+						if subSet === dependencySet {
+							result = (.rejected, subSet)
+						}
+					case .accepted:
+						// Set contains all dependencies, we've got a winner
+						result = (.accepted, nestedResult.dependencySet)
+					}
+				} else {
+					// All done
+					result = (.rejected, dependencySet)
+				}
+			}
+		}
+
+		// By definition result is not nil at this point, so it's ok to force unwrap
+		return result!
```

Comment:
Question: Would asserting in debug and throwing an error in release also do here?